cmd/kubeadm/app/util/etcd/etcd_test.go

```diff
@@ -494,3 +499,294 @@ func TestClient_GetMemberID(t *testing.T) {
+func TestListMembers(t *testing.T) {
+	type fields struct {
+		Endpoints       []string
+		newEtcdClient   func(endpoints []string) (etcdClient, error)
+		listMembersFunc func(backoff *wait.Backoff) (*clientv3.MemberListResponse, error)
+	}
+	tests := []struct {
+		name      string
+		fields    fields
+		want      []Member
+		wantError bool
+	}{
+		{
+			name: "PeerURLs are empty",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{}
+					return f, nil
+				},
+			},
+			want: []Member{},
+		},
+		{
+			name: "PeerURLs are non-empty",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+								},
+							},
+							{
+								ID:   2,
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			want: []Member{
+				{
+					Name:    "member1",
+					PeerURL: "https://member1:2380",
+				},
+				{
+					Name:    "member2",
+					PeerURL: "https://member2:2380",
+				},
+			},
+		},
+		{
+			name: "PeerURLs has multiple urls",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			want: []Member{
+				{
+					Name:    "member1",
+					PeerURL: "https://member1:2380",
+				},
+			},
+		},
+		{
+			name: "ListMembers return error",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+				listMembersFunc: func(backoff *wait.Backoff) (*clientv3.MemberListResponse, error) {
+					return nil, errNotImplemented
+				},
+			},
+			want:      nil,
+			wantError: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Client{
+				Endpoints:       tt.fields.Endpoints,
+				newEtcdClient:   tt.fields.newEtcdClient,
+				listMembersFunc: tt.fields.listMembersFunc,
+			}
+			if c.listMembersFunc == nil {
+				c.listMembersFunc = func(backoff *wait.Backoff) (*clientv3.MemberListResponse, error) {
+					f, _ := c.newEtcdClient([]string{})
+					resp, _ := f.MemberList(context.TODO())
+					return resp, nil
+				}
+			}
```

Comment:
```suggestion
			c := &Client{
				Endpoints:       tt.fields.Endpoints,
				newEtcdClient:   tt.fields.newEtcdClient,
				listMembersFunc: tt.fields.listMembersFunc,
			}
			if c.listMembersFunc == nil {
				c.listMembersFunc = func(backoff *wait.Backoff) (*clientv3.MemberListResponse, error) {
					return c.listMembers(&wait.Backoff{Steps:  1})
				}
			}
```

this is what i meant in earlier comments.
- functions like listMembers can have a custom backoff parameter e.g with one step for unit testing. this means it will exit quickly and not retry:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/wait/backoff.go#L459-L469
- during unit testing of TestListMembers we should use the actual function listMembers(), because that is what we are unit testing...but use the custom backoff
- all other unit test functions TestIsLearner should use a fake listMembersFunc

Reply:
Got it, thanks